service/history/replication/sync_state_retriever.go

```diff
@@ -0,0 +1,409 @@
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/locks"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/namespace"
+	"go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/common/persistence/versionhistory"
+	"go.temporal.io/server/service/history/consts"
+	"go.temporal.io/server/service/history/hsm"
+	"go.temporal.io/server/service/history/shard"
+	"go.temporal.io/server/service/history/workflow"
+	wcache "go.temporal.io/server/service/history/workflow/cache"
+	"google.golang.org/protobuf/proto"
+)
+
+const (
+	Mutation ResultType = iota
+	Snapshot
+	defaultPageSize = 32
+)
+
+type (
+	ResultType      int
+	SyncStateResult struct {
+		Type                       ResultType
+		Mutation                   *persistencepb.WorkflowMutableStateMutation
+		Snapshot                   *persistencepb.WorkflowMutableState
+		EventBlobs                 []*commonpb.DataBlob
+		NewRunInfo                 *replicationpb.NewRunInfo
+		VersionedTransitionHistory []*persistencepb.VersionedTransition
+		LastVersionHistory         *history.VersionHistory
+	}
+	SyncStateRetriever interface {
+		GetSyncWorkflowStateArtifact(
+			ctx context.Context,
+			namespaceID string,
+			execution *commonpb.WorkflowExecution,
+			versionedTransition *persistencepb.VersionedTransition,
+			versionHistories *history.VersionHistories,
+		) (*SyncStateResult, error)
+	}
+
+	SyncStateRetrieverImpl struct {
+		shardContext  shard.Context
+		workflowCache wcache.Cache
+		logger        log.Logger
+	}
+	lastUpdatedStateTransitionGetter interface {
+		GetLastUpdateVersionedTransition() *persistencepb.VersionedTransition
+	}
+)
+
+func NewSyncStateRetriever(
+	shardContext shard.Context,
+	workflowCache wcache.Cache,
+	logger log.Logger,
+) *SyncStateRetrieverImpl {
+	return &SyncStateRetrieverImpl{
+		shardContext:  shardContext,
+		workflowCache: workflowCache,
+		logger:        logger,
+	}
+}
+
+func (s *SyncStateRetrieverImpl) GetSyncWorkflowStateArtifact(
+	ctx context.Context,
+	namespaceID string,
+	execution *commonpb.WorkflowExecution,
+	versionedTransition *persistencepb.VersionedTransition,
+	versionHistories *history.VersionHistories,
+) (_ *SyncStateResult, retError error) {
+	wfCtx, releaseFunc, err := s.workflowCache.GetOrCreateWorkflowExecution(
+		ctx,
+		s.shardContext,
+		namespace.ID(namespaceID),
+		execution,
+		locks.PriorityLow,
+	)
+	defer func() {
+		if releaseFunc != nil {
+			releaseFunc(retError)
+		}
+	}()
+
+	if err != nil {
+		return nil, err
+	}
+	mu, err := wfCtx.LoadMutableState(ctx, s.shardContext)
+	if err != nil {
+		return nil, err
+	}
+	isSameBranch := false
+	if versionedTransition != nil {
+		err = workflow.TransitionHistoryStalenessCheck(mu.GetExecutionInfo().TransitionHistory, versionedTransition)
+		switch {
+		case err == nil:
+			isSameBranch = true
+		case errors.Is(err, consts.ErrStaleState):
+			return nil, serviceerror.NewInvalidArgument(fmt.Sprintf(
```

Comment:
With current replication stack, it will not cause task to be DLQed. It will retry until the it hit ShardOwnership lost and the stream will be reconnected to the correct shard. Shall we return `ShardOwnership Lost` error here?

Reply:
hmm no, unless consistencyCheck is used which performs a reload. Otherwise, this shard could still be owner and has a stale state.

Reply:
Checked the `consistencyCheck`, seems it needs to pass a vector clock to check the consistency, while there is no way for us to get the vector clock besides shardContext, and the consistency check seems will not work in this case.

I changed the error to internal error, so the caller can retry in this case.

Reply:
With the `GetWorkflowLeaseWithConsistencyCheck` method, you can pass nil for the clock and the logic will fallback to perform validation using the supplied validation function.

Reply:
> internal error

internal error is non-retryable. If you want caller to retry, use unavailable.

Reply:
> you can pass nil for the clock
If this is the case, I think it is better to use concistencyCheck. Let me update it.